### DISCUSSION TITLE
Terraformy w devops - jakie taski, jak ze storage dla tfstate i po co plan?

### DISCUSSION THREAD

**Discussion by maciej-skorupka** [AzureCommunityPL/group#245] [Pomoc] [answered]
Hej,w

Jestem fanem azure devops, ale mam wyzwanie z tym, co myśleć o budowaniu środowiska terraformem przy użyciu azure devops. W związku z tym kilka pytań:

1. Jakich tasków używacie dla budowania środowiska terraformem? https://marketplace.visualstudio.com/items?itemName=charleszipp.azure-pipelines-tasks-terraform czy może raczej https://marketplace.visualstudio.com/items?itemName=petergroenewegen.PeterGroenewegen-Xpirit-Vsts-Release-Terraform Pytam w szczególności w kontekście tego czy któreś potrafią w jednym tasku zrobić init i apply.
2. Trzymacie tfstate w storage gdy robicie pipeline? Czy raczej tworzycie per pipeline budujący środowisko i nie trzymacie go gdzieś do współdzielenia?
3. Wrzucacie gdzieś plan, zatwierdzacie go jakoś i potem dopiero apply? Czy może od razu apply bez plan?

**Comment by miqm** [comment]
Pytanie na marginesie (i z ciekawości): czemu terraform?

**Reply by maciej-skorupka** [reply]
Tu nie mam bardzo konkretnej odpowiedzi. Gdzieś między chęcią narzędzia niezależnego od providera, potencjałem porównywania planów z rzeczywistością dzięki tfstate a subiektywnym większym poczuciem czytelności HCLi.

**Reply by miqm** [reply]
W teorii ARM też ma porównanie stanu (what-if), ale niestety to baaardzo słabo działa. Z drugiej strony tfstate to jest wyobrażenie terraforma co do stanu Azure i jakakolwiek ręczna zmiana w portalu spowoduje „rozjazd”.
A co do niezależnego providera to w mojej opinii nie ma czegoś takiego - korzystając z danej chmury jesteś z definicji uzależniony od danego dostawcy. Nie da się tak po prostu przenieść środowiska między AWS a Azure jeśli wykorzystujesz elementy specyficzne dla danej platformy, zawsze będzie się to wiązać z mniejszym lub większym przepisaniem IaaC.
Ponadto TF dla Azure jest zbudowany w oparciu o SDK for Go. I dość często zdarza się, że kończysz z pisaniem ARMa i uruchamianiem go przez TF, bo nie ma „natywnej” funkcji w TF.

**Reply by smereczynski** [reply]
Ja tylko dodam, że TF nie jest narzędziem niezależnym od dostawcy platformy cloud, bo musisz ogarnąć providera, czyli ARM, od którego jest zależny i to ma miejsce w przypadku każdego kolejnego dostawcy. Niezależna jest co najwyżej binarka i semantyka całości składni, bo poszczególne elementy zależą od providera. Dyskusja (bardziej filozoficzna) powinna być na temat: Czy TF upraszcza cokolwiek, czy komplikuje? :)

**Reply by miqm** [reply]
>Dyskusja (bardziej filozoficzna) powinna być na temat: Czy TF upraszcza cokolwiek, czy komplikuje? :)

Brzmi jak idealny temat na konferencję :)

**Comment by smajchrz** [comment]
Ad 2,3

Tutaj jest to fajnie rozpisane tyle ze dla Hashi Cloud i GitHub'a. Plan leci do PR a przy merge jest apply. State oczywiście w blob.

https://learn.hashicorp.com/tutorials/terraform/github-actions?in=terraform/automation

**Comment by piotrgwiazda** [comment, marked as answer]
Hej


Tu opisane podejście ogólne (autoreklama)
https://blog.gft.com/pl/2020/03/04/secure-terraform-delivery-pipeline-best-practices-part-1-2/

Odpowiada też na pytanie "dlaczego Terraform?".

Tu natomiast szczegółowy przykład implementacji w Azure DevOps
https://blog.gft.com/pl/2020/04/24/automating-infrastructure-deployment-on-azure-using-ci-cd-pipeline-and-terraform/

- Wydaje mi się, że wykorzystywałem te taski "charles zipp".
- Plan wymaga zatwierdzenia, a dodatkowo powinien przejść weryfjkację pod
kątem zgodności z politykami (Terraform Sentinel, Checkov itp)
- Więcej o state file w artykule

niedz., 14 mar 2021, 02:38 użytkownik Sebastian Majchrzak <
***@***.***> napisał:

**Reply by maciej-skorupka** [reply]
Bardzo konkretne, dzięki. Przetrawię:)

**Comment by kamil-mrzyglod** [comment]
Terraform w ADO czy ogólnie organizacji to trochę szerszy temat. Ja osobiście głównie używałem https://marketplace.visualstudio.com/items?itemName=ms-devlabs.custom-terraform-tasks, ale na samym końcu którego taska użyjesz nie ma takiego dużego znaczenia. TF możesz odpalać za pomocą dowolnego taska CMD i często okazuje się, że to jest najlepsza opcja.

Jak sobie taki pipeline skonfigurujesz zależy od miejsca składowania plików infrastruktury i ogólnego design'u Twojej aplikacji. Jeśli przykładowo masz centralne repo z modułami, które jest współdzielone pomiędzy zespołami, to może się okazać, że domyślne taski nie spełnią swojej roli bo masz potrzebę dodatkowego uwierzytelnienia. 

Nie wiem trochę co masz na myśli mówiąc o współdzieleniu pliku stanu, ale ja z definicji traktuję ten plik jako sekret i nie ma żadnej mowy o współdzieleniu. Może on być wtedy składowany w Storage Account na oddzielnej RG albo nawet subskrypcji, do której wjazd ma tylko i wyłącznie principal, który odpala pipeline. Generalna zasada - używasz Azure RM backendu, nie trzymasz pliku stanu na agencie ;)

Spokojnie możesz robić apply bez żadnego approve'a, ale jak Ci TF zrobi recreate na wszystkich VM bo "gdzieś się zmienił jeden parametr, który to wymusił", to możesz mieć pretensje tylko do siebie. Prosty setup:

- Stage A - validate & plan
- Stage B - manual approval
- Stage C - apply

Robisz wtedy `terraform plan` z parametrem `-out` aby gdzieś wypluć plan, manual approval służy potwierdzeniu poprawności planu i w Stage C wykorzystujesz zmaterializowany plan aby robotę wykonać zgodnie z założeniami.

**Reply by maciej-skorupka** [reply]
Mówiąc o współdzieleniu miałem na myśli właśnie tfstate w storage, ale może rzeczywiście raczej powinienem myśleć o używaniu tego pliku tam tylko przez jedno coś. Bo dalej nie do końca rozumiem czemu w zasadzie jest tfstate w takim miejscu, skoro i tak terraform plan porównuje stan. Robię tak, ale nie do końca rozumiem konieczność.

Co do manualnego potwierdzania planu to właśnie na taki case liczyłem. Dzięki, bo czuję się przekonany:)

**Reply by miqm** [reply]
Jeśli masz jakieś hasła, to one są w tfstate zapisane w formie czystej: https://github.com/hashicorp/terraform/issues/516 i dlatego ten plik trzeba pilnować :)

**Reply by kamil-mrzyglod** [reply]
@maciej-skorupka TF, w przeciwieństwie do ARM, przerzuca na Ciebie odpowiedzialność za utrzymanie pliku stanu i ostatecznie jest to chyba najbardziej dyskusyjny feature z powodu różnych głupot z tym związanych. Koncepcyjnie to wygląda fajnie, możesz umieścić plik stanu w dowolnym miejscu, które jest współdzielone, dzięki czemu w łatwy sposób możesz inkrementować zmiany z dowolnego miejsca w dowolnie dużym zespole. W rzeczywistości to jest największa dziura w całym narzędziu i potencjalny wektor ataku.  Jak nie zadbasz o odpowiednie zabezpieczenie tego pliku to InfoSec zrobi Ci z tyłka jesień średniowiecza :)

**Reply by KamilWiecek** [reply]
> Spokojnie możesz robić apply bez żadnego approve'a, ale jak Ci TF zrobi recreate na wszystkich VM bo "gdzieś się zmienił jeden parametr, który to wymusił", to możesz mieć pretensje tylko do siebie.

Dokładnie, trzeba być bardzo bardzo uważnym. Atrybuty takie jak description grupy w Azure AD, lub zmiana procesu w Azure DevOps spowodują utworzenie resource na nowo! Na szczęście jest to dość dobrze udokumentowane. Możesz jeszcze skorzystać z lifecycle:

```terraform
  lifecycle {
    prevent_destroy = true
    ignore_changes  = [
       ...
    ]
  }
```

**Comment by KamilWiecek** [comment]
3. W jednym z projektów wrzucamy plan jako komentarz do PR. To nasz dodatkowy "check".

Proces wygląda tak:

![image](https://user-images.githubusercontent.com/57139732/111082518-07621380-8509-11eb-8d6e-e48a13ef79af.png)

Komentarz wygląda tak:

![image](https://user-images.githubusercontent.com/57139732/111082354-18f6eb80-8508-11eb-966a-232d81deec5a.png)

Wykorzystujemy do tego build validation. Pipeline jest uniwersalny: jeżli odpalany jest jako validation w PR to wrzuca komentarz. Jeżeli uruchaminy jest po merge to tworzy infrastrukturę. Branch polisy wymusza rozwiązanie wszystkich komentarzy. 

![image](https://user-images.githubusercontent.com/57139732/111082587-5f991580-8509-11eb-9ef0-b5f8ac42b90b.png)

[Opisałem to tutaj](https://www.kamilwiecek.com/blog/azure-dev-sec-platform-ops ).
